addon/-private/wrap-descriptor.js

```diff
@@ -0,0 +1,100 @@
+import { run as runValidator } from './validators';
+
+/**
+ * Transforms a `field` descriptor into a property `getter/setter` that
+ * ensures that the value matches the validators.
+ *
+ * @param {Object} element
+ * @param {Function} validator
+ */
+function wrapFieldDescriptor(
+  { descriptor, initializer, key, placement },
+  validator
+) {
+  const { ...descriptorToApply } = descriptor;
+  delete descriptorToApply.writable;
+
+  let initialized = false;
+  let cachedValue;
+
+  const newElement = {
+    key,
+    placement,
+    kind: 'method',
+    descriptor: {
+      ...descriptorToApply,
+      get() {
+        if (!initialized) {
+          initialized = true;
+          cachedValue = initializer ? initializer.call(this) : undefined;
+        }
+
+        return cachedValue;
+      },
+      set(newValue) {
+        initialized = true;
+
+        runValidator(validator, this.constructor, key, newValue, 'set');
```

Comment:
One thing I'm realizing is that unless we call `super`, the last class that defines the argument will always win:

```js
class Foo {
  @argument('string') baz;
}

class Bar {
  @argument('number') baz;
}
```

We may need to store all of the validators in an array with `addValidationFor` and then run all of them in the setter.

Reply:
I'm not sure I'm following -- Are the usage in `Foo` and `Bar` related, even though the classes are independent?

Reply:
yes sorry, that was a typo, corrected

Reply:
Gotcha. We'd talked about this briefly earlier -- we have a few options for defining arguments for the same property on multiple components in a hierarchy

1. First one (lowest in the chain) is run
2. Last one (highest in the chain) is run
3. All of them are run

We currently do the first option there, but I'm happy to refactor to run all of them. That should be pretty easy now with our `and` combinator!